torch/fx/experimental/validator.py

```diff
@@ -0,0 +1,419 @@
+
+        def add_source_expr(self, e: z3.BoolRef) -> None:
+            self._source_exprs.add(e)
+
+        def add_target_expr(self, e: sympy.Expr) -> None:
+            self._check_freesymbols(e)
+            self._target_exprs.add(self.to_z3_boolean_expr(e))
+
+        def add_assertion(self, e: Union[z3.BoolRef, sympy.Basic]) -> None:
+            if isinstance(e, sympy.Basic):
+                self._check_freesymbols(e)
+                ref = self.to_z3_boolean_expr(e)
+            else:
+                ref = e
+            assert isinstance(ref, z3.BoolRef)
+            self._assertions.add(ref)
+
+        # The result of a validation run.
+        @dataclass
+        class Result:
+            success: bool
+
+            # Mapping of the name of each free variable to the value assigned to it.
+            model: Optional[z3.ModelRef] = None
+
+            # List of the source expressions that failed due to the assignment.
+            failed_source_expr: Optional[List[z3.BoolRef]] = None
+
+            # List of the target expressions.
+            target_exprs: Optional[Set[z3.BoolRef]] = None
+
+        def validate(self) -> "TranslationValidator.Result":
+            from torch._dynamo.utils import dynamo_timed
+
+            # Here, we use "QF_NRA" logic for the solver, since guards have no quantifiers
+            # and are potentially non-linear.
+            solver = z3.SolverFor("QF_NRA")
+
+            # Add all the assertions to the solver.
+            for assertion in self._assertions:
+                solver.add(assertion)
+
+            # "Is there any case where it's TRUE for the target expressions,
+            #  but FALSE for the source expressions?"
+            solver.add(z3.Not(z3.And(*self._source_exprs)))
+            solver.add(*self._target_exprs)
+
+            log.debug("translation validation: start")
+            r = dynamo_timed()(solver.check)()
+            if r == z3.sat:
+                # Target expressions are unsound.
+                # Log the found model and the source expressions that failed.
+                model = solver.model()
+                return self.Result(
+                    success=False,
+                    model=model,
+                    failed_source_expr=[inp for inp in self._source_exprs if not model.evaluate(inp)],
+                    target_exprs=self._target_exprs,
+                )
+            else:
+                if r == z3.unknown:
+                    # Could not find a solution. It didn't fail, but it also
+                    # didn't succeed. Canceling the validation execution (keyboard
+                    # interrupt) also gets to this branch.
+                    log.warning("translation validation: could not validate")
+                else:
+                    # Target expressions are sound.
+                    assert r == z3.unsat
+                    log.debug("translation validation: success")
+                return self.Result(success=True)
```

Comment:
On second thoughts, I don't think I should be doing this, but fail on `z3.unknown`.

Reply:
@ezyang 
This makes lots of tests fail. What do you think?
Maybe we could land it as is, and try to tweak the validator later: e.g. find a better solver, use only reals...

Reply:
Yeah, let's work on it more later

Reply:
Maybe @nunoplopes has some ideas

Reply:
You should ignore z3.unknown. That is either because the solver gave up or you reached a time/memory out, for example.
Tweaking the encoding and solver options is a whole different game. I would leave for a follow up patch. I can help with that ofc.
But I would try switching everything into bit-vectors; that's usually the most performing option.

Reply:
Got it. So, for now, I will revert to the old behavior: issue a warning.